node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,999 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The Network Bridge Subsystem - protocol multiplexer for Polkadot.
+
+use parity_scale_codec::{Encode, Decode};
+use futures::prelude::*;
+use futures::future::BoxFuture;
+use futures::stream::BoxStream;
+
+use sc_network::{
+	ObservedRole, ReputationChange, PeerId,
+	Event as NetworkEvent,
+};
+use sp_runtime::ConsensusEngineId;
+
+use polkadot_subsystem::{
+	FromOverseer, OverseerSignal, Subsystem, SubsystemContext, SpawnedSubsystem, SubsystemError,
+	SubsystemResult,
+};
+use polkadot_subsystem::messages::{NetworkBridgeEvent, NetworkBridgeMessage, AllMessages};
+use node_primitives::{ProtocolId, View};
+use polkadot_primitives::{Block, Hash};
+
+use std::collections::btree_map::{BTreeMap, Entry as BEntry};
+use std::collections::hash_map::{HashMap, Entry as HEntry};
+use std::pin::Pin;
+use std::sync::Arc;
+
+const MAX_VIEW_HEADS: usize = 5;
+
+/// The engine ID of the polkadot network protocol.
+pub const POLKADOT_ENGINE_ID: ConsensusEngineId = *b"dot2";
+/// The protocol name.
+pub const POLKADOT_PROTOCOL_NAME: &[u8] = b"/polkadot/2";
+
+const MALFORMED_MESSAGE_COST: ReputationChange
+	= ReputationChange::new(-500, "Malformed Network-bridge message");
+const UNKNOWN_PROTO_COST: ReputationChange
+	= ReputationChange::new(-50, "Message sent to unknown protocol");
+const MALFORMED_VIEW_COST: ReputationChange
+	= ReputationChange::new(-500, "Malformed view");
+
+/// Messages received on the network.
+#[derive(Debug, Encode, Decode, Clone)]
+pub enum WireMessage {
+	/// A message from a peer on a specific protocol.
+	#[codec(index = "1")]
+	ProtocolMessage(ProtocolId, Vec<u8>),
+	/// A view update from a peer.
+	#[codec(index = "2")]
+	ViewUpdate(View),
+}
+
+/// Information about the notifications protocol. Should be used during network configuration
+/// or shortly after startup to register the protocol with the network service.
+pub fn notifications_protocol_info() -> (ConsensusEngineId, std::borrow::Cow<'static, [u8]>) {
+	(POLKADOT_ENGINE_ID, POLKADOT_PROTOCOL_NAME.into())
+}
+
+/// An action to be carried out by the network.
+#[derive(PartialEq)]
+pub enum NetworkAction {
+	/// Note a change in reputation for a peer.
+	ReputationChange(PeerId, ReputationChange),
+	/// Write a notification to a given peer.
+	WriteNotification(PeerId, Vec<u8>),
+}
+
+/// An abstraction over networking for the purposes of this subsystem.
+pub trait Network: Clone + Send + 'static {
+	/// Get a stream of all events occurring on the network. This may include events unrelated
+	/// to the Polkadot protocol - the user of this function should filter only for events related
+	/// to the [`POLKADOT_ENGINE_ID`](POLKADOT_ENGINE_ID).
+	fn event_stream(&mut self) -> BoxStream<'static, NetworkEvent>;
+
+	/// Get access to an underlying sink for all network actions.
+	fn action_sink<'a>(&'a mut self) -> Pin<
```

Comment:
I don't understand what the benefit of this method is. Why does this trait offer two ways of changing the reputation and writing a notification?

As this method does not add any additional functionality, why not add the logic below as a trait default implementation either on this trait directly, or as an extension trait (like `StreamExt`, `FutureExt`, ...)?

Reply:
We want to have access to a full stream because it lets us do `send_all` and other similar things. The functions `report_peer` and `write_notification` are trait defaults, but convenient. so a trait implementation needs only provide the `action_sink` function. Making an extension trait is kind of overkill here